src/libopenrave-core/jsonparser/jsondownloader.cpp

```diff
@@ -0,0 +1,330 @@
+        return;
+    }
+
+    const uint64_t startTimestampUS = utils::GetMonotonicTime();
+    int numDownloads = 0;
+    while (!_mapDownloadContexts.empty()) {
+        // check if any handle still running
+        int numRunningHandles = 0;
+        const CURLMcode performCode = curl_multi_perform(_curlm, &numRunningHandles);
+        if (performCode) {
+            throw OPENRAVE_EXCEPTION_FORMAT("failed to download, curl_multi_perform() failed with code %d", (int)performCode, ORE_InvalidArguments);
+        }
+        RAVELOG_VERBOSE_FORMAT("curl_multi_perform(): numRunningHandles = %d", numRunningHandles);
+        if (numRunningHandles > 0) {
+            // poll for 100ms
+            const CURLMcode pollCode = curl_multi_poll(_curlm, NULL, 0, 100, NULL);
+            if (pollCode) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download, curl_multi_poll() failed with code %d", (int)pollCode, ORE_InvalidArguments);
+            }
+        }
+
+        // check for timeout
+        const uint64_t currentTimestampUS = utils::GetMonotonicTime();
+        if (currentTimestampUS > startTimestampUS + timeoutUS) {
+            throw OPENRAVE_EXCEPTION_FORMAT("timed out waiting for download to finish, timeout is %d[us]", timeoutUS, ORE_Timeout);
+        }
+
+        // process all messages
+        for (;;) {
+            int numMessagesInQueue = 0;
+            const CURLMsg *curlMessage = curl_multi_info_read(_curlm, &numMessagesInQueue);
+            RAVELOG_VERBOSE_FORMAT("curl_multi_info_read(): numMessagesInQueue = %d", numMessagesInQueue);
+            if (!curlMessage) {
+                break;
+            }
+            RAVELOG_VERBOSE_FORMAT("curl_multi_info_read(): curlMessage->msg = %d", (int)curlMessage->msg);
+            if (curlMessage->msg != CURLMSG_DONE) {
+                continue;
+            }
+
+            // remove handle
+            const CURLMcode removeHandleCode = curl_multi_remove_handle(_curlm, curlMessage->easy_handle);
+            if (removeHandleCode) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download, curl_multi_remove_handle() failed with code %d", (int)removeHandleCode, ORE_InvalidArguments);
+            }
+            const std::map<CURL*, JSONDownloadContextPtr>::iterator it = _mapDownloadContexts.find(curlMessage->easy_handle);
+            if (it == _mapDownloadContexts.end()) {
+                throw OPENRAVE_EXCEPTION_FORMAT0("curl download finished, but failed to find corresponding context, cannot continue", ORE_InvalidArguments);
+            }
+            JSONDownloadContextPtr pContext = it->second;
+            _mapDownloadContexts.erase(it);
+
+            // check result
+            const CURLcode curlCode = curlMessage->data.result;
+            if (curlCode != CURLE_OK) {
+                throw OPENRAVE_EXCEPTION_FORMAT("failed to download uri \"%s\": %s", pContext->uri%curl_easy_strerror(curlCode), ORE_InvalidURI);
```

Comment:
ORE_CurlInvalidResponse

Reply:
Yes, because the handle in curlMessage was created by curl_easy_init